<a href="https://colab.research.google.com/github/lkostenko/python_da_homeworks/blob/main/hw_3_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
data_path = '/content/drive/MyDrive/курс Аналітик/навчання/Модуль 3. Робота з табличними даними: Pandas та Numpy/data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [ ]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [ ]:
grouped = df.groupby('Branch')['gross income'].sum().sort_values(ascending=False)
grouped

Branch
C    5265.1765
A    5057.1605
B    5057.0320
Name: gross income, dtype: float64

In [ ]:
print(f"Філіал {grouped.idxmax()} має найвищий прибуток, {grouped.max()} ")

Філіал C має найвищий прибуток, 5265.1765 


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [ ]:
 df.groupby('City')['gross income'].sum().sort_values(ascending=False)

City
Naypyitaw    5265.1765
Yangon       5057.1605
Mandalay     5057.0320
Name: gross income, dtype: float64

In [ ]:
df.pivot_table(index='Branch', columns='City', values='gross income', aggfunc='sum', fill_value = 0)

City    Mandalay  Naypyitaw     Yangon
Branch                                
A          0.000     0.0000  5057.1605
B       5057.032     0.0000     0.0000
C          0.000  5265.1765     0.0000

З даної таблиці можна побачити, що найбільше знчення в місті Naypyitaw - 5265.1765
Я вважаю, що можна розмісти ще один сагазин в даному місті.


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [ ]:
invoices_by_category = df.pivot_table(index='Branch', columns= "Product line", values = "Invoice ID", aggfunc="count")
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

In [ ]:
max_prodline_branch = invoices_by_category.idxmax()
max_prodline_invoices = invoices_by_category.max()

max_prodline_branch_invoices = pd.DataFrame({
    "Branch": max_prodline_branch,
    "Invoices count": max_prodline_invoices
})

max_prodline_branch_invoices

Branch  Invoices count
Product line                                 
Electronic accessories      A              60
Fashion accessories         C              65
Food and beverages          C              66
Health and beauty           B              53
Home and lifestyle          A              65
Sports and travel           B              62

Дана таблиця нам показує філіал, в якому найбільше інвойсів за певною категорією твоарів

In [ ]:
max_branch_prodline = invoices_by_category.idxmax(axis=1)
max_branch_invoices = invoices_by_category.max(axis=1)

max_branch_prodline_invoices = pd.DataFrame({
    "Product line": max_branch_prodline,
    "Invoices count": max_branch_invoices
})
max_branch_prodline_invoices

Product line  Invoices count
Branch                                     
A        Home and lifestyle              65
B       Fashion accessories              62
C        Food and beverages              66

Дана таблиця нам показує, яка категорія товарів найпопулярніша в певному філіалі та кількість інвойсів по цій категорії.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [ ]:
invoices_by_category["Electronic accessories"].idxmax()

'A'

In [ ]:
print(f'У філіалі {invoices_by_category["Electronic accessories"].idxmax()} - {invoices_by_category["Electronic accessories"].max()} інвойсів із покупкою товарів категорії "Електронні аксесуари"')

У філіалі A - 60 інвойсів із покупкою товарів категорії "Електронні аксесуари"


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**1.Чи є різниця в популярності продуктів серед різних типів клієнтів (члени та звичайні покупці) в різних містах. Та чи впливає гендер на тип продукту в різних містах.**

В таблиці city_product_customer можна побачити скільки продуктів кожного типу було придбано різними категоріми клієнтів - Member та Normal. Також можна відслідкувати, які типи продуктів більше купували жінки, а які чоловіки в певних містах.

In [ ]:
city_product_customer = pd.pivot_table(
    df,
    values='Quantity',
    index=['City', 'Product line'],
    columns=['Customer type','Gender'],
    aggfunc='sum'
)
city_product_customer

Customer type                    Member      Normal     
Gender                           Female Male Female Male
City      Product line                                  
Mandalay  Electronic accessories     76   71     80   89
          Fashion accessories        79   64     98   56
          Food and beverages        117   43     52   58
          Health and beauty          39  133     73   75
          Home and lifestyle         76   61     71   87
          Sports and travel          92   73     58   99
Naypyitaw Electronic accessories     61   59    103  110
          Fashion accessories        98   65     94   85
          Food and beverages        127   66    104   72
          Health and beauty          58   69     62   88
          Home and lifestyle        110   33     42   60
          Sports and travel          94   57     96   18
Yangon    Electronic accessories     71   91     97   63
          Fashion accessories        80   53     81   49
          Food and beverages         42  111     72   88
          Health and beauty          61   68     50   78
          Home and lifestyle        105  105     94   67
          Sports and travel         106   71     50  106

Наступна таблиця показує, які категорії товарів зустрічаються найчастіше в різних містах для різних типів клієнтів

In [ ]:
def get_mode(series):
  return series.mode().iloc[0]

In [ ]:
pivot_table = pd.pivot_table(
    df,
    values='Product line',
    index=['Customer type', 'Gender'],
    columns='City',
    aggfunc=get_mode,
)
pivot_table

City                             Mandalay               Naypyitaw  \
Customer type Gender                                                
Member        Female   Food and beverages      Food and beverages   
              Male      Health and beauty       Health and beauty   
Normal        Female  Fashion accessories  Electronic accessories   
              Male      Sports and travel     Fashion accessories   

City                                  Yangon  
Customer type Gender                          
Member        Female       Sports and travel  
              Male    Electronic accessories  
Normal        Female  Electronic accessories  
              Male         Sports and travel

Наступна таблиця показує які категорії товарів купували найчатіше в різних містах, різні типи покупців (групувала по Quantity). Дані відрізняються від попередньї таблиці, бо в колонці 'Product line' категорія товарів зазначена лише 1 раз, а в Quantity рахується кількість товарів, як були куплені в даній категорії

In [ ]:
def get_max_product(group):
  quantity_prodline = group.groupby('Product line')['Quantity'].sum()
  return quantity_prodline.idxmax()

In [ ]:
pivot_table_product = df.groupby(['Customer type', 'Gender', 'City']).apply(get_max_product).unstack()
pivot_table_product

<ipython-input-8-2771ad9a1f64>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  pivot_table_product = df.groupby(['Customer type', 'Gender', 'City']).apply(get_max_product).unstack()


City                             Mandalay               Naypyitaw  \
Customer type Gender                                                
Member        Female   Food and beverages      Food and beverages   
              Male      Health and beauty       Health and beauty   
Normal        Female  Fashion accessories      Food and beverages   
              Male      Sports and travel  Electronic accessories   

City                                  Yangon  
Customer type Gender                          
Member        Female       Sports and travel  
              Male        Food and beverages  
Normal        Female  Electronic accessories  
              Male         Sports and travel

В наступній таблиці я порахувала кількість продуктів (по категоріям), які буди куплені найчастіше

In [ ]:
def get_max_quantity(group):
  quantity_prodline = group.groupby('Product line')['Quantity'].sum()
  return quantity_prodline.max()

In [ ]:
pivot_table_quantity = df.groupby(['Customer type', 'Gender', 'City']).apply(get_max_quantity).unstack()
pivot_table_quantity

<ipython-input-10-7b1ae15306e8>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  pivot_table_quantity = df.groupby(['Customer type', 'Gender', 'City']).apply(get_max_quantity).unstack()


City                  Mandalay  Naypyitaw  Yangon
Customer type Gender                             
Member        Female       117        127     106
              Male         133         69     111
Normal        Female        98        104      97
              Male          99        110     106

Ця таблиця показує об'єднані значення

In [ ]:
concatenated_df = pd.concat([pivot_table_product, pivot_table_quantity], keys=['Категорія товарів', 'Кількість товарів'], axis=1)
concatenated_df

Категорія товарів                          \
City                             Mandalay               Naypyitaw   
Customer type Gender                                                
Member        Female   Food and beverages      Food and beverages   
              Male      Health and beauty       Health and beauty   
Normal        Female  Fashion accessories      Food and beverages   
              Male      Sports and travel  Electronic accessories   

                                             Кількість товарів            \
City                                  Yangon          Mandalay Naypyitaw   
Customer type Gender                                                       
Member        Female       Sports and travel               117       127   
              Male        Food and beverages               133        69   
Normal        Female  Electronic accessories                98       104   
              Male         Sports and travel                99       110   

                             
City                 Yangon  
Customer type Gender         
Member        Female    106  
              Male      111  
Normal        Female     97  
              Male      106

Тут також об'єднані значення, але в іншому форматі.

In [ ]:
def get_max_product_and_quantity(group):
    quantity_prodline = group.groupby('Product line')['Quantity'].sum()
    max_product = quantity_prodline.idxmax()
    max_quantity = quantity_prodline.max()
    return pd.Series([max_product, max_quantity], index=['Категорія товарів', 'Кількість товарів'])

In [ ]:
pivot_table_product_quantity = df.groupby(['Customer type', 'Gender', 'City']).apply(get_max_product_and_quantity)
pivot_table_product_quantity

<ipython-input-18-c354bad4d9b4>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  pivot_table_product_quantity = df.groupby(['Customer type', 'Gender', 'City']).apply(get_max_product_and_quantity)


Категорія товарів  Кількість товарів
Customer type Gender City                                                
Member        Female Mandalay       Food and beverages                117
                     Naypyitaw      Food and beverages                127
                     Yangon          Sports and travel                106
              Male   Mandalay        Health and beauty                133
                     Naypyitaw       Health and beauty                 69
                     Yangon         Food and beverages                111
Normal        Female Mandalay      Fashion accessories                 98
                     Naypyitaw      Food and beverages                104
                     Yangon     Electronic accessories                 97
              Male   Mandalay        Sports and travel                 99
                     Naypyitaw  Electronic accessories                110
                     Yangon          Sports and travel                106

**Висновки:**

Food and beverages є дуже популярною категорією товарів у Mandalay (для жінок з типом "Member") та Naypyitaw (для жінок з типом "Member" і "Normal")

Sports and travel — популярна категорія для чоловіків у Mandalay та Yangon, зокрема для "Normal" клієнтів

Health and beauty користується популярністю у Mandalay серед чоловіків, але в інших містах її попит не настільки високий

У Yangon різноманітність категорій товарів відображається більш виражено: "Sports and travel" для жінок з типом "Member", а також Food and beverages для чоловіків з типом "Member"


**За містами:**

У Naypyitaw і Mandalay клієнти обирають більше Food and beverages, тому можна звернути увагу на популяризацію цієї категорії в рекламних кампаніях у цих містах.

Yangon демонструє більш розподілений попит між різними категоріями, що свідчить про великий попит на різноманітні товари.

**За типом клієнта:**

Для чоловіків з типом клієнта "Member" і "Normal" у різних містах популярні Health and beauty та Sports and travel, що може вказувати на зацікавленість в активному способі життя, здоров'ї та відпочинку.

Жінки з типом клієнта "Member" у Naypyitaw і Mandalay часто вибирають Food and beverages, що може свідчити про високий попит на продукти харчування і напої.

*Рекомендації*

Mandalay - фокус на Food and beverages для жінок і Health and beauty для чоловіків. Це може включати рекламні кампанії або спеціальні акції для цих груп клієнтів.

Naypyitaw - Food and beverages популярна серед обох типів клієнтів.ТОму можна збільшити постачання / урізноманітнити асортимент.

Yangon - розглянути збільшення асортименту у Sports and travel та Food and beverages, оскільки ці категорії товарів популярні серед клієнтів в різних сегментах.

Стратегія для Member може включати бонуси та лояльні програми, особливо для Food and beverages і Sports and travel.
Для Normal клієнтів в різних містах можна застосувати індивідуальні пропозиції.


**2.Чи є певні місяці або дні тижня, коли клієнти купують більше товарів, і чи це пов'язано з сезоном або святами**

In [5]:
def get_date(date):
  return int(date.split('/')[1])

In [6]:
def get_month(date):
  return int(date.split('/')[0])

In [7]:
df['Day'] = df.Date.apply(get_date)
df['Month'] = df.Date.apply(get_month)

In [ ]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total  ...  \
0       Health and beauty       74.69         7  26.1415  548.9715  ...   
1  Electronic accessories       15.28         5   3.8200   80.2200  ...   
2      Home and lifestyle       46.33         7  16.2155  340.5255  ...   
3       Health and beauty       58.22         8  23.2880  489.0480  ...   
4       Sports and travel       86.31         7  30.2085  634.3785  ...   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

   (Day, Month)  Day  Month  Year  
0             1    5      1  2019  
1             3    8      3  2019  
2             3    3      3  2019  
3             1   27      1  2019  
4             2    8      2  2019  

[5 rows x 21 columns]

In [47]:
day_quantity_pivot_table = pd.pivot_table(
    df,
    values='Quantity',
    index=['Month', 'Day'],
    columns= ['City', 'Product line'],
    aggfunc='sum'
)
day_quantity_pivot_table.fillna(0)

City                       Mandalay                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
Month Day                                                                    
1     1                         6.0                 0.0               18.0   
      2                         6.0                 9.0                0.0   
      3                         0.0                 0.0                0.0   
      4                         1.0                 4.0                0.0   
      5                         0.0                 1.0                1.0   
...                             ...                 ...                ...   
3     26                        3.0                 0.0                0.0   
      27                        2.0                 0.0                6.0   
      28                        0.0                 3.0                0.0   
      29                        0.0                 0.0                0.0   
      30                        0.0                 0.0                2.0   

City                                                                 \
Product line Health and beauty Home and lifestyle Sports and travel   
Month Day                                                             
1     1                    0.0                0.0               0.0   
      2                   10.0                0.0               6.0   
      3                    8.0                6.0               1.0   
      4                    0.0                0.0               3.0   
      5                    0.0                9.0               0.0   
...                        ...                ...               ...   
3     26                   0.0                9.0               1.0   
      27                   0.0                0.0               6.0   
      28                   0.0                0.0               0.0   
      29                  27.0                0.0               0.0   
      30                  17.0                0.0               0.0   

City                      Naypyitaw                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
Month Day                                                                    
1     1                         2.0                 0.0                0.0   
      2                         0.0                 2.0                0.0   
      3                         0.0                 0.0                0.0   
      4                         9.0                 0.0                0.0   
      5                         0.0                10.0                0.0   
...                             ...                 ...                ...   
3     26                        2.0                 1.0                9.0   
      27                        0.0                 9.0                0.0   
      28                        3.0                 0.0                7.0   
      29                       14.0                 0.0                0.0   
      30                        3.0                 8.0                8.0   

City                                                                 \
Product line Health and beauty Home and lifestyle Sports and travel   
Month Day                                                             
1     1                    2.0                0.0              16.0   
      2                    0.0                0.0               8.0   
      3                    3.0                0.0               0.0   
      4                    0.0                0.0               0.0   
      5                    7.0                0.0               0.0   
...                        ...                ...               ...   
3     26                   2.0                0.0               0.0   
      27                   0.0                0.0               0.0   
      28                   0.0        

In [48]:
city_day = day_quantity_pivot_table.stack().idxmax()
max_value = day_quantity_pivot_table.stack().max()

<ipython-input-48-24bd65766ba5>:1: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  city_day = day_quantity_pivot_table.stack().idxmax()
<ipython-input-48-24bd65766ba5>:2: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  max_value = day_quantity_pivot_table.stack().max()


In [49]:
max_day_quantity = pd.DataFrame({"Max value": max_value, "Month / Day / Product" : city_day})
max_day_quantity

Max value            Month / Day / Product
City                                                 
Mandalay        29.0  (1, 31, Electronic accessories)
Naypyitaw       31.0   (3, 3, Electronic accessories)
Yangon          24.0        (3, 4, Health and beauty)

In [50]:
day_total_pivot_table = pd.pivot_table(
    df,
    values='Total',
    index=['Month', 'Day' ],
    columns= ['City', 'Product line'],
    aggfunc='sum'
)
day_total_pivot_table.fillna(0)

City                       Mandalay                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
Month Day                                                                    
1     1                    470.6730              0.0000          1066.0230   
      2                    138.6630            383.7645             0.0000   
      3                      0.0000              0.0000             0.0000   
      4                     75.7785            288.5820             0.0000   
      5                      0.0000             77.6580            74.7600   
...                             ...                 ...                ...   
3     26                   143.9865              0.0000             0.0000   
      27                   116.9070              0.0000           193.6305   
      28                     0.0000             95.6655             0.0000   
      29                     0.0000              0.0000             0.0000   
      30                     0.0000              0.0000           196.1400   

City                                                                 \
Product line Health and beauty Home and lifestyle Sports and travel   
Month Day                                                             
1     1                  0.000             0.0000             0.000   
      2                359.205             0.0000           281.169   
      3                520.800           397.8450            98.049   
      4                  0.000             0.0000           146.223   
      5                  0.000           334.3410             0.000   
...                        ...                ...               ...   
3     26                 0.000           116.1405            41.076   
      27                 0.000             0.0000           588.357   
      28                 0.000             0.0000             0.000   
      29              2059.932             0.0000             0.000   
      30              1528.317             0.0000             0.000   

City                      Naypyitaw                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
Month Day                                                                    
1     1                    132.7620              0.0000             0.0000   
      2                      0.0000            209.6220             0.0000   
      3                      0.0000              0.0000             0.0000   
      4                    629.8425              0.0000             0.0000   
      5                      0.0000            445.5780             0.0000   
...                             ...                 ...                ...   
3     26                   176.9250             51.1455           154.1295   
      27                     0.0000            943.2990             0.0000   
      28                   150.0975              0.0000           330.7080   
      29                   985.6980              0.0000             0.0000   
      30                   304.9830             85.5120           609.1680   

City                                                                 \
Product line Health and beauty Home and lifestyle Sports and travel   
Month Day                                                             
1     1               132.0270                0.0           572.376   
      2                 0.0000                0.0           266.028   
      3               124.0260                0.0             0.000   
      4                 0.0000                0.0             0.000   
      5               579.8415                0.0             0.000   
...                        ...                ...               ...   
3     26               91.7700                0.0             0.000   
      27                0.0000                0.0             0.000   
      28                0.0000        

In [51]:
city_day_total = day_total_pivot_table.stack().idxmax()
max_value_total = day_total_pivot_table.stack().max()

<ipython-input-51-3a7692c7316d>:1: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  city_day_total = day_total_pivot_table.stack().idxmax()
<ipython-input-51-3a7692c7316d>:2: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  max_value_total = day_total_pivot_table.stack().max()


In [52]:
max_day_total = pd.DataFrame({"Max Total": max_value_total, "Month / Day / Product" : city_day_total})
max_day_total

Max Total           Month / Day / Product
City                                                
Mandalay    2167.599      (1, 28, Health and beauty)
Naypyitaw   1562.085  (3, 3, Electronic accessories)
Yangon      1461.558      (2, 3, Home and lifestyle)

In [53]:
concatenated_df_city_day = pd.concat([max_day_quantity, max_day_total], axis=1)
concatenated_df_city_day

Max value            Month / Day / Product  Max Total  \
City                                                               
Mandalay        29.0  (1, 31, Electronic accessories)   2167.599   
Naypyitaw       31.0   (3, 3, Electronic accessories)   1562.085   
Yangon          24.0        (3, 4, Health and beauty)   1461.558   

                    Month / Day / Product  
City                                       
Mandalay       (1, 28, Health and beauty)  
Naypyitaw  (3, 3, Electronic accessories)  
Yangon         (2, 3, Home and lifestyle)

Висновки:

У Mandalay найбільші обсяги продажів припадають на 31 січня у категорії Electronic accessories, а також 28 січня — у категорії Health and beauty.
У Naypyitaw найбільше продажів зафіксовано 3 березня, знову ж таки в категорії Electronic accessories.
У Yangon найбільше продано 4 березня, з найбільшими продажами у категорії Health and beauty, а також 3 лютого — у категорії Home and lifestyle.

Вищі продажі на кінець місяця або початок місяця можуть бути пов'язані з зарплатними періодами, коли люди мають більше фінансових можливостей для покупок.
Це також може бути пов'язано з сезонними акціями або святами, оскільки деякі дні (наприклад, початок березня) з сезонними знижками.

Оскільки продажі значно зростають в певні дні місяця, варто розглянути можливість запуску маркетингових кампаній або акцій в ці періоди. Якщо пік продажів припадає на кінець місяця, можна зробити спеціальні пропозиції або знижки на продукцію саме в цей період.  Можна запропонувати персоналізовані рекомендації для клієнтів на основі їхньої історії покупок або попиту в цей день (наприклад, пропонувати знижки на Electronic accessories в місті Naypyitaw).
Знаючи найбільш популярні дні та продукти, можна краще планувати запаси товарів, щоб уникнути дефіциту товару в періоди високих продажів.

Також можна перевірити стать та тип клієнта. Це дозволить ще краще налаштувати маркетингові стратегії.

In [60]:
day_quantity_pivot_table_gender = pd.pivot_table(
    df,
    values='Quantity',
    index=['Customer type', 'Gender', 'Month', 'Day'],
    columns= ['City', 'Product line'],
    aggfunc='sum'
)
day_quantity_pivot_table_gender.fillna(0)

City                                         Mandalay                      \
Product line                   Electronic accessories Fashion accessories   
Customer type Gender Month Day                                              
Member        Female 1     1                      0.0                 0.0   
                           2                      0.0                 0.0   
                           3                      0.0                 0.0   
                           4                      0.0                 4.0   
                           5                      0.0                 1.0   
...                                               ...                 ...   
Normal        Male   3     26                     0.0                 0.0   
                           27                     0.0                 0.0   
                           28                     0.0                 0.0   
                           29                     0.0                 0.0   
                           30                     0.0                 0.0   

City                                                                 \
Product line                   Food and beverages Health and beauty   
Customer type Gender Month Day                                        
Member        Female 1     1                 10.0               0.0   
                           2                  0.0               0.0   
                           3                  0.0               0.0   
                           4                  0.0               0.0   
                           5                  0.0               0.0   
...                                           ...               ...   
Normal        Male   3     26                 0.0               0.0   
                           27                 3.0               0.0   
                           28                 0.0               0.0   
                           29                 0.0              17.0   
                           30                 0.0               0.0   

City                                                                 \
Product line                   Home and lifestyle Sports and travel   
Customer type Gender Month Day                                        
Member        Female 1     1                  0.0               0.0   
                           2                  0.0               0.0   
                           3                  0.0               0.0   
                           4                  0.0               0.0   
                           5                  9.0               0.0   
...                                           ...               ...   
Normal        Male   3     26                 0.0               0.0   
                           27                 0.0               6.0   
                           28                 0.0               0.0   
                           29                 0.0               0.0   
                           30                 0.0               0.0   

City                                        Naypyitaw                      \
Product line                   Electronic accessories Fashion accessories   
Customer type Gender Month Day                                              
Member        Female 1     1                      0.0                 0.0   
                           2                      0.0                 0.0   
                           3                      0.0                 0.0   
                           4                      9.0                 0.0   
                           5                      0.0                 1.0   
...                                               ...                 ...   
Normal        Male   3     26                     0.0                 0.0   
                           27                     0.0                 0.0   
                           28                     0.0                 0.0   
       

In [64]:
city_day_gender = day_quantity_pivot_table_gender.stack(future_stack=True).idxmax()
max_value_gender = day_quantity_pivot_table_gender.stack(future_stack=True).max()

max_day_quantity_gender_gender = pd.DataFrame({"Max value": max_value_gender, "Customer Type / Gender / Month / Day / Product" : city_day_gender})
max_day_quantity_gender_gender

Max value Customer Type / Gender / Month / Day / Product
City                                                               
Mandalay        21.0      (Member, Female, 3, 5, Health and beauty)
Naypyitaw       18.0    (Normal, Female, 1, 16, Food and beverages)
Yangon          17.0   (Member, Female, 2, 10, Fashion accessories)

In [65]:
day_total_pivot_table_gender = pd.pivot_table(
    df,
    values='Total',
    index=['Customer type', 'Gender', 'Month', 'Day' ],
    columns= ['City', 'Product line'],
    aggfunc='sum'
)
day_total_pivot_table_gender.fillna(0)

City                                         Mandalay                      \
Product line                   Electronic accessories Fashion accessories   
Customer type Gender Month Day                                              
Member        Female 1     1                      0.0               0.000   
                           2                      0.0               0.000   
                           3                      0.0               0.000   
                           4                      0.0             288.582   
                           5                      0.0              77.658   
...                                               ...                 ...   
Normal        Male   3     26                     0.0               0.000   
                           27                     0.0               0.000   
                           28                     0.0               0.000   
                           29                     0.0               0.000   
                           30                     0.0               0.000   

City                                                                 \
Product line                   Food and beverages Health and beauty   
Customer type Gender Month Day                                        
Member        Female 1     1              888.615            0.0000   
                           2                0.000            0.0000   
                           3                0.000            0.0000   
                           4                0.000            0.0000   
                           5                0.000            0.0000   
...                                           ...               ...   
Normal        Male   3     26               0.000            0.0000   
                           27             101.808            0.0000   
                           28               0.000            0.0000   
                           29               0.000         1450.1445   
                           30               0.000            0.0000   

City                                                                 \
Product line                   Home and lifestyle Sports and travel   
Customer type Gender Month Day                                        
Member        Female 1     1                0.000             0.000   
                           2                0.000             0.000   
                           3                0.000             0.000   
                           4                0.000             0.000   
                           5              334.341             0.000   
...                                           ...               ...   
Normal        Male   3     26               0.000             0.000   
                           27               0.000           588.357   
                           28               0.000             0.000   
                           29               0.000             0.000   
                           30               0.000             0.000   

City                                        Naypyitaw                      \
Product line                   Electronic accessories Fashion accessories   
Customer type Gender Month Day                                              
Member        Female 1     1                   0.0000               0.000   
                           2                   0.0000               0.000   
                           3                   0.0000               0.000   
                           4                 629.8425               0.000   
                           5                   0.0000              33.495   
...                                               ...                 ...   
Normal        Male   3     26                  0.0000               0.000   
                           27                  0.0000               0.000   
                           28                  0.0000               0.000   
       

In [68]:

city_day_total_gender = day_total_pivot_table_gender.stack(future_stack=True).idxmax()
max_value_total_gender = day_total_pivot_table_gender.stack(future_stack=True).max()


max_day_total_gender = pd.DataFrame({"Max Total": max_value_total_gender, "Customer Type / Gender / Month / Day / Product" : city_day_total_gender})
max_day_total_gender

Max Total Customer Type / Gender / Month / Day / Product
City                                                               
Mandalay    1528.317       (Member, Male, 3, 30, Health and beauty)
Naypyitaw   1378.377    (Normal, Female, 1, 16, Food and beverages)
Yangon      1139.544        (Member, Male, 3, 4, Health and beauty)

In [69]:
concatenated_df_city_day_gender = pd.concat([max_day_quantity_gender_gender, max_day_total_gender], axis=1)
concatenated_df_city_day_gender

Max value Customer Type / Gender / Month / Day / Product  \
City                                                                  
Mandalay        21.0      (Member, Female, 3, 5, Health and beauty)   
Naypyitaw       18.0    (Normal, Female, 1, 16, Food and beverages)   
Yangon          17.0   (Member, Female, 2, 10, Fashion accessories)   

           Max Total Customer Type / Gender / Month / Day / Product  
City                                                                 
Mandalay    1528.317       (Member, Male, 3, 30, Health and beauty)  
Naypyitaw   1378.377    (Normal, Female, 1, 16, Food and beverages)  
Yangon      1139.544        (Member, Male, 3, 4, Health and beauty)

У всіх трьох містах максимальні обсяги продажів відбуваються серед Member  — це означає, що лояльні покупці можуть здійснювати більші покупки.
Жінки є ключовими покупцями, оскільки в усіх випадках максимальні продажі були здійснені серед жінок.
Березень, здається, є пікованим місяцем для продажів, тому варто зосередити маркетингові кампанії на цей період. Можна провести сезонні розпродажі або спеціальні акції для мумберів в цьому місяці, щоб максимізувати продажі.



**3. Чи є зв'язок між методом оплати і типом клієнта?**

In [89]:
payment_by_customer = df.groupby(['Customer type', 'Payment', 'City','Product line'])['Total'].agg(['sum', 'count'])
payment_by_customer.head(10)

sum  count
Customer type Payment City      Product line                            
Member        Cash    Mandalay  Electronic accessories  3113.5965     13
                                Fashion accessories     2187.5175     11
                                Food and beverages      1237.6455      4
                                Health and beauty       3442.2150      6
                                Home and lifestyle      2971.3635      7
                                Sports and travel       3962.1435     12
                      Naypyitaw Electronic accessories  3244.0275      7
                                Fashion accessories     3156.8040     10
                                Food and beverages      5868.7545     18
                                Health and beauty       3304.5495      8

In [105]:
most_frequent_payment = payment_by_customer.groupby(['Customer type', 'City','Product line' ])['count'].idxmax()
most_frequent_payment_info = payment_by_customer.loc[most_frequent_payment]
most_frequent_payment_info

sum  count
Customer type Payment     City      Product line                            
Member        Cash        Mandalay  Electronic accessories  3113.5965     13
                                    Fashion accessories     2187.5175     11
              Credit card Mandalay  Food and beverages      5797.3650     13
                                    Health and beauty       4643.9925     12
              Cash        Mandalay  Home and lifestyle      2971.3635      7
                                    Sports and travel       3962.1435     12
              Ewallet     Naypyitaw Electronic accessories  3058.2195      8
              Credit card Naypyitaw Fashion accessories     4281.7320     11
              Cash        Naypyitaw Food and beverages      5868.7545     18
              Credit card Naypyitaw Health and beauty       3623.3400     10
                                    Home and lifestyle      2926.9485     10
                                    Sports and travel       4629.2400     15
              Cash        Yangon    Electronic accessories  2047.8780     11
              Ewallet     Yangon    Fashion accessories     3939.9570     12
              Cash        Yangon    Food and beverages      3651.5955     11
                                    Health and beauty       2234.9775     10
                                    Home and lifestyle      6208.6395     14
              Ewallet     Yangon    Sports and travel       4034.5620     12
Normal        Cash        Mandalay  Electronic accessories  3802.9320     11
                                    Fashion accessories     3469.6935     11
              Credit card Mandalay  Food and beverages      2777.7960     12
              Cash        Mandalay  Health and beauty       4285.7640     11
              Ewallet     Mandalay  Home and lifestyle      5308.9575     15
              Cash        Mandalay  Sports and travel       3979.2900     14
                          Naypyitaw Electronic accessories  6402.9315     22
                                    Fashion accessories     4512.2700     12
                                    Food and beverages      5479.4670     13
              Ewallet     Naypyitaw Health and beauty       3688.4820     11
                                    Home and lifestyle      3405.3705      7
                                    Sports and travel       3866.3205      7
              Credit card Yangon    Electronic accessories  4433.4885     12
              Cash        Yangon    Fashion accessories     3175.5360     11
              Credit card Yangon    Food and beverages      3118.3425     11
                                    Health and beauty       2029.6080      9
              Ewallet     Yangon    Home and lifestyle      3939.7050     14
              Cash        Yangon    Sports and travel       3823.8480     12

Cash є найбільш популярним методом оплати серед різних типів клієнтів у різних містах для багатьох категорій продуктів (особливо для Food and beverages та Electronic accessories).
Credit card використовується переважно для більш дорогих покупок,у таких категоріях, як Food and beverages або Health and beauty, в залежності від міста.
Ewallet найменш популярний.

Мембери зазвичай розраховуються кешем, рідше картою і найрідше Ewallet. У normal покупців ситуація схожа.

Оскільки Cash залишається найбільш популярним методом оплати, можна зосередитися на оптимізації платіжних систем, що підтримують готівкові розрахунки (наприклад, зручні термінали або кешбек-акці
